File: aip_protocol/mesh.py
"""
AIP Mesh Client — Connects local RevocationStore to the cloud mesh.

Usage:
    from aip_protocol.mesh import MeshClient

    mesh = MeshClient("mesh_xxx_your_api_key")
    mesh.connect()

    # Now your local revocation store auto-updates in real-time.
    # When someone revokes an agent via the mesh API, your local
    # store gets the update within <50ms.

    # You can also revoke from code:
    mesh.revoke("did:web:acme.com:agents:rogue-bot", reason="gone_rogue")
    mesh.suspend("did:web:acme.com:agents:suspicious", duration=1800)

    # Check status:
    status = mesh.status("did:web:acme.com:agents:my-bot")
"""

from __future__ import annotations

import json
import threading
import time
from datetime import datetime, timezone
from typing import Callable
from urllib.request import Request, urlopen
from urllib.error import URLError

from aip_protocol.revocation import RevocationStore


DEFAULT_MESH_URL = "https://mesh.synthexai.tech"


class MeshClient:
    """
    Connects a local RevocationStore to the AIP Revocation Mesh.

    On connect(), starts a background thread that receives real-time
    revocation events and applies them to your local store.

    API calls (revoke, suspend, status) go through REST.
    Real-time updates come through SSE (Server-Sent Events).
    """

    def __init__(
        self,
        api_key: str,
        mesh_url: str = DEFAULT_MESH_URL,
        store: RevocationStore | None = None,
        on_revocation: Callable[[dict], None] | None = None,
    ):
        self.api_key = api_key
        self.mesh_url = mesh_url.rstrip("/")
        self.store = store or RevocationStore()
        self.on_revocation = on_revocation
        self._connected = False
        self._thread: threading.Thread | None = None
        self._stop = threading.Event()

    def connect(self) -> "MeshClient":
        """Start receiving real-time revocation events from the mesh."""
        if self._connected:
            return self

        self._stop.clear()
        self._thread = threading.Thread(
            target=self._sse_loop,
            daemon=True,
            name="aip-mesh-client",
        )
        self._thread.start()
        self._connected = True
        return self

    def disconnect(self):
        """Stop receiving events."""
        self._stop.set()
        self._connected = False

    def _sse_loop(self):
        """Background SSE listener with auto-reconnect."""
        while not self._stop.is_set():
            try:
                req = Request(
                    f"{self.mesh_url}/mesh/events",
                    headers={
                        "Authorization": f"Bearer {self.api_key}",
                        "Accept": "text/event-stream",
                    },
                )
                with urlopen(req, timeout=60) as resp:
                    for line in resp:
                        if self._stop.is_set():
                            break
                        line = line.decode("utf-8").strip()
                        if line.startswith("data: "):
                            try:
                                event = json.loads(line[6:])
                                self._handle_event(event)
                            except json.JSONDecodeError:
                                pass
            except (URLError, OSError, TimeoutError):
                # Reconnect after backoff
                if not self._stop.is_set():
                    self._stop.wait(5.0)

    def _handle_event(self, event: dict):
        """Apply a mesh event to the local RevocationStore."""
        event_type = event.get("type", "")
        agent_id = event.get("agent_id", "")

        if event_type == "revocation":
            self.store.revoke(
                agent_id=agent_id,
                reason=event.get("reason", "mesh_revocation"),
                revoked_by=event.get("revoked_by", "mesh"),
                scope=event.get("scope", "global"),
            )
        elif event_type == "suspension":
            suspended_until = event.get("suspended_until", "")
            if suspended_until:
                until_dt = datetime.fromisoformat(suspended_until)
                now = datetime.now(timezone.utc)
                duration = max(1, int((until_dt - now).total_seconds()))
                self.store.suspend(
                    agent_id=agent_id,
                    duration_seconds=duration,
                    reason=event.get("reason", "mesh_suspension"),
                )
        elif event_type == "reinstatement":
            self.store.reinstate(agent_id)

        # Call user callback if set
        if self.on_revocation:
            try:
                self.on_revocation(event)
            except Exception:
                pass

    # ── REST API calls ─────────────────────────────────────────────

    def _api_call(self, endpoint: str, data: dict) -> dict:
        """Make an authenticated API call to the mesh."""
        body = json.dumps(data).encode("utf-8")
        req = Request(
            f"{self.mesh_url}{endpoint}",
            data=body,
            headers={
                "Authorization": f"Bearer {self.api_key}",
                "Content-Type": "application/json",
            },
            method="POST",
        )
        with urlopen(req, timeout=10) as resp:
            return json.loads(resp.read())

    def revoke(self, agent_id: str, reason: str = "manual_revocation", scope: str = "global") -> dict:
        """Revoke an agent via the mesh (broadcasts to all connected clients)."""
        result = self._api_call("/mesh/revoke", {
            "agent_id": agent_id,
            "reason": reason,
            "scope": scope,
        })
        # Also apply locally immediately
        self.store.revoke(agent_id, reason, "mesh", scope)
        return result

    def suspend(self, agent_id: str, duration: int = 1800, reason: str = "auto_suspend") -> dict:
        """Suspend an agent temporarily via the mesh."""
        result = self._api_call("/mesh/suspend", {
            "agent_id": agent_id,
            "duration_seconds": duration,
            "reason": reason,
        })
        self.store.suspend(agent_id, duration, reason)
        return result

    def status(self, agent_id: str) -> dict:
        """Check an agent's revocation status on the mesh."""
        return self._api_call("/mesh/status", {"agent_id": agent_id})

    def reinstate(self, agent_id: str) -> dict:
        """Reinstate a revoked/suspended agent via the mesh."""
        result = self._api_call("/mesh/reinstate", {"agent_id": agent_id})
        self.store.reinstate(agent_id)
        return result

    def register_agent(self, agent_id: str) -> dict:
        """Register an agent on the mesh (counts against tier limit)."""
        return self._api_call("/mesh/agents/register", {"agent_id": agent_id})

    def usage(self) -> dict:
        """Get current month's usage and tier info."""
        req = Request(
            f"{self.mesh_url}/mesh/usage",
            headers={"Authorization": f"Bearer {self.api_key}"},
        )
        with urlopen(req, timeout=10) as resp:
            return json.loads(resp.read())

    @property
    def connected(self) -> bool:
        return self._connected

File: sdks/typescript/src/mesh.ts
/**
 * AIP Mesh Client — TypeScript
 *
 * Connects to the AIP Revocation Mesh for real-time kill switch.
 *
 * Usage:
 *   import { MeshClient } from "aip-protocol/mesh";
 *
 *   const mesh = new MeshClient("mesh_xxx_your_api_key");
 *   mesh.connect();
 *
 *   // Revoke an agent across all connected services:
 *   await mesh.revoke("did:web:acme.com:agents:rogue-bot", "gone_rogue");
 *
 *   // Check status:
 *   const status = await mesh.status("did:web:acme.com:agents:my-bot");
 */

import { RevocationStore } from "./revocation.js";

const DEFAULT_MESH_URL = "https://mesh.synthexai.tech";

export interface MeshEvent {
  type: "revocation" | "suspension" | "reinstatement";
  id: string;
  agent_id: string;
  action: string;
  reason?: string;
  scope?: string;
  suspended_until?: string;
  timestamp: string;
}

export interface MeshRevokeResponse {
  id: string;
  agent_id: string;
  action: string;
  reason: string;
  propagated_to: number;
  propagation_ms: number;
  timestamp: string;
}

export interface MeshStatusResponse {
  agent_id: string;
  status: "active" | "revoked" | "suspended";
  reason?: string;
  revoked_at?: string;
  suspended_until?: string;
}

export class MeshClient {
  private apiKey: string;
  private meshUrl: string;
  private store: RevocationStore;
  private onRevocation?: (event: MeshEvent) => void;
  private _connected = false;
  private abortController?: AbortController;

  constructor(
    apiKey: string,
    options: {
      meshUrl?: string;
      store?: RevocationStore;
      onRevocation?: (event: MeshEvent) => void;
    } = {}
  ) {
    this.apiKey = apiKey;
    this.meshUrl = (options.meshUrl || DEFAULT_MESH_URL).replace(/\/$/, "");
    this.store = options.store || new RevocationStore();
    this.onRevocation = options.onRevocation;
  }

  /**
   * Start receiving real-time revocation events from the mesh.
   */
  async connect(): Promise<this> {
    if (this._connected) return this;

    this.abortController = new AbortController();
    this._connected = true;

    // Start SSE listener in background
    this.sseLoop();

    return this;
  }

  /**
   * Stop receiving events.
   */
  disconnect(): void {
    this.abortController?.abort();
    this._connected = false;
  }

  private async sseLoop(): Promise<void> {
    while (this._connected) {
      try {
        const response = await fetch(`${this.meshUrl}/mesh/events`, {
          headers: {
            Authorization: `Bearer ${this.apiKey}`,
            Accept: "text/event-stream",
          },
          signal: this.abortController?.signal,
        });

        if (!response.ok || !response.body) {
          throw new Error(`SSE connection failed: ${response.status}`);
        }

        const reader = response.body.getReader();
        const decoder = new TextDecoder();
        let buffer = "";

        while (this._connected) {
          const { done, value } = await reader.read();
          if (done) break;

          buffer += decoder.decode(value, { stream: true });
          const lines = buffer.split("\n");
          buffer = lines.pop() || "";

          for (const line of lines) {
            const trimmed = line.trim();
            if (trimmed.startsWith("data: ")) {
              try {
                const event: MeshEvent = JSON.parse(trimmed.slice(6));
                this.handleEvent(event);
              } catch {
                // Skip malformed events
              }
            }
          }
        }
      } catch (err) {
        if (!this._connected) break;
        // Reconnect after backoff
        await new Promise((r) => setTimeout(r, 5000));
      }
    }
  }

  private handleEvent(event: MeshEvent): void {
    if (event.type === "revocation") {
      this.store.revoke(
        event.agent_id,
        event.reason || "mesh_revocation",
        "mesh",
        event.scope || "global"
      );
    } else if (event.type === "suspension") {
      if (event.suspended_until) {
        const until = new Date(event.suspended_until);
        const now = new Date();
        const duration = Math.max(1, Math.floor((until.getTime() - now.getTime()) / 1000));
        this.store.suspend(event.agent_id, duration, event.reason || "mesh_suspension");
      }
    } else if (event.type === "reinstatement") {
      // RevocationStore doesn't have reinstate — just clear by re-creating
      // In a full impl, we'd add reinstate() to the store
    }

    this.onRevocation?.(event);
  }

  // ── REST API calls ──────────────────────────────────────────────

  private async apiCall<T>(endpoint: string, data: Record<string, unknown>): Promise<T> {
    const resp = await fetch(`${this.meshUrl}${endpoint}`, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${this.apiKey}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify(data),
    });
    if (!resp.ok) {
      const error = await resp.text();
      throw new Error(`Mesh API error (${resp.status}): ${error}`);
    }
    return resp.json() as Promise<T>;
  }

  /**
   * Revoke an agent via the mesh (broadcasts to all connected clients).
   */
  async revoke(
    agentId: string,
    reason = "manual_revocation",
    scope = "global"
  ): Promise<MeshRevokeResponse> {
    const result = await this.apiCall<MeshRevokeResponse>("/mesh/revoke", {
      agent_id: agentId,
      reason,
      scope,
    });
    this.store.revoke(agentId, reason, "mesh", scope);
    return result;
  }

  /**
   * Suspend an agent temporarily via the mesh.
   */
  async suspend(
    agentId: string,
    durationSeconds = 1800,
    reason = "auto_suspend"
  ): Promise<MeshRevokeResponse> {
    const result = await this.apiCall<MeshRevokeResponse>("/mesh/suspend", {
      agent_id: agentId,
      duration_seconds: durationSeconds,
      reason,
    });
    this.store.suspend(agentId, durationSeconds, reason);
    return result;
  }

  /**
   * Check an agent's revocation status on the mesh.
   */
  async status(agentId: string): Promise<MeshStatusResponse> {
    return this.apiCall<MeshStatusResponse>("/mesh/status", {
      agent_id: agentId,
    });
  }

  /**
   * Reinstate a revoked/suspended agent via the mesh.
   */
  async reinstate(agentId: string): Promise<{ agent_id: string; status: string }> {
    return this.apiCall("/mesh/reinstate", { agent_id: agentId });
  }

  /**
   * Register an agent on the mesh (counts against tier limit).
   */
  async registerAgent(agentId: string): Promise<Record<string, unknown>> {
    return this.apiCall("/mesh/agents/register", { agent_id: agentId });
  }

  /**
   * Get current month's usage and tier info.
   */
  async usage(): Promise<Record<string, unknown>> {
    const resp = await fetch(`${this.meshUrl}/mesh/usage`, {
      headers: { Authorization: `Bearer ${this.apiKey}` },
    });
    return resp.json();
  }

  get connected(): boolean {
    return this._connected;
  }
}
